pylele_solid.py

```diff
@@ -0,0 +1,149 @@
+#!/usr/bin/env python3
+
+"""
+    Abstract Base and Concrete Classes for all Gugulele Solid objects
+"""
+
+from __future__ import annotations
+
+import os
+import sys
+import argparse
+import datetime
+from pathlib import Path
+from abc import ABC, abstractmethod
+
+from pylele_api import ShapeAPI, Shape
+from pylele_config import Fidelity, Implementation, CARBON, GRAY, LITE_GRAY,DARK_GRAY, ORANGE, WHITE
+
+COLORS = {
+    'carbon': CARBON, 
+    'gray': GRAY, 
+    'lite_gray' : LITE_GRAY,
+    'dark_gray' : DARK_GRAY, 
+    'orange' : ORANGE, 
+    'white': WHITE
+}
+
+def make_or_exist_path(out_path):
+    """ Check a directory exist, and generate if not """
+
+    if not os.path.isdir(out_path):
+        # Path.mkdir(out_path)
+        os.makedirs(out_path)
+
+    assert os.path.isdir(out_path), f"Cannot export to non directory: {out_path}"
+
+def lele_solid_parser(parser=None):
+    """
+    LeleSolid Command Line Interface
+    """
+    if parser is None:
+        parser = argparse.ArgumentParser(description='LeleSolid Configuration')
+
+    ## options ######################################################
+    parser.add_argument("-i", "--implementation", help="Underlying engine implementation, default cadquery",
+                        type=Implementation, choices=list(Implementation), default=Implementation.BLENDER)
+    parser.add_argument("-f", "--fidelity", help="Mesh fidelity for smoothness, default low",
+                        type=Fidelity, choices=list(Fidelity), default=Fidelity.LOW)
+    parser.add_argument("-color", "--color", help="This solid is a cutting.", type=str,
+                        choices = COLORS.keys(), default = COLORS['white']
+                        )
+    parser.add_argument("-outdir", "--outdir", help="Output directory.", type=str,default='build')
+    parser.add_argument("-is_cut", "--is_cut",
+                    help="This solid is a cutting.",
+                    action='store_true')
+
+    return parser
+
+class LeleSolid(ABC):
```

Comment:
I like where this is going. I was thinking similar ideas of needing another abstraction layer between LelePart and ShapeAPI + Shape.  If this class is introduced, then we should move some code from LelePart here to avoid duplicates. Would you like to take a first cut? 

But I am still not sure of how or whether we need another main like entry point distinct from pylele_main(). Let's discuss...

Reply:
Yes, exactly.
Even though I love your work about ukulele, I think you started something that potentially is even more valuable to a wider audience: an implementation-independent python wrapper of a solid modeling library.
This could be used to generate any design, and therefor, can possibly be split into a separate library on which pylele (that should stay an ukulele-focused project) will depend, but potentially other projects too. 

> then we should move some code from LelePart here to avoid duplicates. Would you like to take a first cut?

I can.
But perhaps in order to keep a working repository I may start duplicating code.
I would start building from the ground up the different elements of pylele.
Once we are confident we can remove the older implementation and continue development on the newer hierarchycal codebase.

> But I am still not sure of how or whether we need another main like entry point distinct from pylele_main(). Let's discuss...

Having different main() is actually one of my objectives: as you know I am interested in developing independent parts (for instance a stand alone fretboard, or neck, or brdige, etc, ...).
My vision is that of having each component to have its own simplified parser, and all the parsers contributes to the main parser that generates a full ukulele.
A similar implementation would also reduce the annoyance of a slow implementation (openscad ?), because you will not need to run a full ukulele generation, if you only are intereste in a bridge, for instance.